### PR TITLE
Initial commit for support of Type 3 and Type 4 LISP packets.

### PR DESCRIPTION
**Pull request by ritsz**
Currently works for Only type 3 and 4 LISP packets and only IPv4 and IPv6 AFI. 
Handles multiple EID record as well.
21:43:33.889543 IP 192.168.0.105.4342 > localhost.4342:  LISP Map-Notify EID 10.30.1.100/32 LOC 20.20.8.251
EID 10.30.1.102/32 LOC 20.20.8.255
EID 10.30.1.98/32 LOC 20.20.8.254

working on for support for LCAF AFI and multiple LOC entries.

### CODE REVIEW

**Comment by guyharris** [comment]
As mcr said in his e-mail reply to your e-mail about this, "You'll want to add your test case to tests/TESTLIST, and provide the expected output as well."  "Add your test case" also includes checking in the pcap file for the test case.

**Comment by guyharris** [comment]
(BTW, whose bright idea was it to name the protocol after a rather well-known, but unrelated, programming language? :-))

**Comment by fxlb** [comment]
Please, could you fix the warnings?
To get them do : `touch .devel print-lisp.c; ./configure; make -s`
```
./print-lisp.c:54:2: warning: initialization discards ‘const’ qualifier from pointer target type [enabled by default]

./print-lisp.c:55:2: warning: initialization discards ‘const’ qualifier from pointer target type [enabled by default]

./print-lisp.c:56:2: warning: initialization discards ‘const’ qualifier from pointer target type [enabled by default]

./print-lisp.c:57:2: warning: initialization discards ‘const’ qualifier from pointer target type [enabled by default]

./print-lisp.c:58:2: warning: initialization discards ‘const’ qualifier from pointer target type [enabled by default]

./print-lisp.c:59:2: warning: initialization discards ‘const’ qualifier from pointer target type [enabled by default]

./print-lisp.c:99:9: warning: no previous prototype for ‘lisp_type’ [-Wmissing-prototypes]

./print-lisp.c: In function ‘lisp_print’:

./print-lisp.c:119:8: warning: cast discards ‘__attribute__((const))’ qualifier from pointer target type [-Wcast-qual]

./print-lisp.c:136:4: warning: implicit declaration of function ‘getname’ [-Wimplicit-function-declaration]

./print-lisp.c:136:4: warning: format ‘%s’ expects argument of type ‘char *’, but argument 3 has type ‘int’ [-Wformat]

./print-lisp.c:142:4: warning: implicit declaration of function ‘getname6’ [-Wimplicit-function-declaration]

./print-lisp.c:142:4: warning: format ‘%s’ expects argument of type ‘char *’, but argument 3 has type ‘int’ [-Wformat]

./print-lisp.c:157:28: warning: cast discards ‘__attribute__((const))’ qualifier from pointer target type [-Wcast-qual]

./print-lisp.c:163:4: warning: format ‘%s’ expects argument of type ‘char *’, but argument 3 has type ‘int’ [-Wformat]

./print-lisp.c:168:4: warning: format ‘%s’ expects argument of type ‘char *’, but argument 3 has type ‘int’ [-Wformat]

./print-lisp.c:108:23: warning: unused parameter ‘is_multi’ [-Wunused-parameter]
```

**Comment by ritsz** [comment]
@fxlb Thanks, fixed the build warnings
@guyharris Added the testcases, pcaps and output files. Tests are passing locally for me. About the naming though, you'll have to take that up with Dino Farinacci I guess :D

**Comment by infrastation** [comment]
Please do not assume the decoder has a complete packet in the provided buffer, the frame generally may be truncated at any random position. Check the `TTEST` and `TCHECK` macros in other decoders to see how it should work.

**Comment by ritsz** [comment]
Quite a few fields in the LISP packet are variable fields. Handling the TCHECK for them accordingly.

**Comment by ritsz** [comment]
Could someone take a look at this PR please? I have handled the issues that have been brought up so far. Tests, TCHECK and build warnings have all been taken care of as far as I can see. Is there anything else that I need to do?

**Comment by fxlb** [comment]
After a second review:
- `s/LISP_PORT/LISP_CONTROL_PORT/` (Protocol/Iana define : `lisp-data   4341  udp  LISP Data Packets` and `lisp-control   4342  udp   LISP Control Packets`) ;
- in `print-udp.c`, put your test on LISP_CONTROL_PORT after the one on GENEVE_PORT ;
- In verbose mode (-v), the dissector should print all fields of the packets (try your pcap tests files in wireshark) ;
- The test cases are only for IPv4 and not showing loops (several records, several locators, ...) ;
- Some variable names are too cryptic (type_PSIR, reserved_ETamM, ...) ;
- in structure definitions : `s/<4_spaces>/<tab>/` ;
- For better parsing : `s/"Control Message"/"Control-Message"/` (no space as other texts).

**Comment by fxlb** [comment]
In `6.1.1.  LISP Packet Type Allocations`, the value for `LISP Encapsulated Control Message` is `8` not `5`

**Comment by ritsz** [comment]
@fxlb Thanks for reviewing, fixed the highlighted issues in the latest commit. The verbose prints all the major fields in the headers. Added test cases for multiple records and IPv6 as well.
Are there any other changes required?

**Comment by infrastation** [comment]
It would be helpful to mention RFC 7215 in the change to `udp.h`.

**Comment by fxlb** [comment]
or RFC6830 ?

**Comment by ritsz** [comment]
@infrastation @fxlb The LISP control plane RFC is 6830. Updated the header file with the RFC information. Any further changes?

**Comment by fxlb** [comment]
Thanks for the updates.
I need a little time for a new review.

File: print-lisp.c [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+	"Map-Request",
+	"Map-Reply",
+	"Map-Register",
+	"Map-Notify",
+	"Unallocated-type",
+	"Unallocated-type",
+	"Unallocated-type",
+	"Control-Message",
+};
+
```
Comment (fxlb) [inline print-lisp.c]: A better way:
```
#define LISP_MAP_REQUEST            1
#define LISP_MAP_REPLY              2
#define LISP_MAP_REGISTER           3
#define LISP_MAP_NOTIFY             4
#define LISP_ENCAPS_CONTROL_MESSAGE 8
static const struct tok type_name[] = {
        { 0,                           "Reserved"        },
        { LISP_MAP_REQUEST,            "Map-Request"     },
        { LISP_MAP_REPLY,              "Map-Reply"       },
        { LISP_MAP_REGISTER,           "Map-Register"    },
        { LISP_MAP_NOTIFY,             "Map-Notify"      },
        { LISP_ENCAPS_CONTROL_MESSAGE, "Control-Message" },
        { 0, NULL }
};
```
With this tok structure, you can print the `type_name` with possible error with `tok2str` function:
`tok2str(type_name, "unknown-type-%u", type)`

File: print-lisp.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+} UNALIGNED lisp_map_register_hdr;
+
+#define MAP_REGISTER_HDR_LEN sizeof(lisp_map_register_hdr)
+
+static const char *auth_type[] = {
+	"None",
+	"SHA1",
+	"SHA256",
+};
+
```
Comment (fxlb) [inline print-lisp.c]: A tok structure and tok2str() usage is better

File: print-lisp.c [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+			/*
+			 * No support for LCAF right now.
+			 */
+			return;
+			break;
+		}
+		ND_PRINT((ndo, " maps to"));
+		if (loc_count > 1)
+			ND_PRINT((ndo, " %u Locators,", loc_count));
+
```
Comment (fxlb) [inline print-lisp.c]: No need for " maps to", just ", %u locator(s), "

File: print-lisp.c [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+static void action_flag(netdissect_options *ndo, uint8_t act_auth_inc_res);
+
+
+void lisp_print(netdissect_options *ndo, const u_char *bp, u_int length)
+{
+	/* Check if enough bytes for header are available */
+	ND_TCHECK2(*bp, MAP_REGISTER_HDR_LEN);
+	const lisp_map_register_hdr *lisp_hdr = (const lisp_map_register_hdr *) bp;
+	uint8_t type = lisp_type(lisp_hdr);
+	ND_PRINT((ndo, ":LISP %s,", type_name[type]));
```
Comment (fxlb) [inline print-lisp.c]: s/:LISP/LISP/

File: print-lisp.c [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+				#ifdef INET6
+				ND_PRINT((ndo, " LOC %s", getname6(ndo, loc_ip_pointer)));
+				#endif
+				packet_offset += 16;
+				break;
+			default:
+				break;
+			}
+			if (ndo->ndo_vflag) {
+				ND_PRINT((ndo, "(Priority/Weight %u/%u, "
```
Comment (fxlb) [inline print-lisp.c]: Need a space before "(Priority"

File: print-lisp.c [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+typedef struct map_register_hdr {
+	uint8_t type_and_flag;
+	uint8_t reserved;
+	uint8_t reserved_and_flag2;
+	uint8_t record_count;
+	uint64_t nonce;
+	uint16_t key_id;
+	uint16_t auth_data_len;
+	uint8_t auth_data[0];
```
Comment (fxlb) [inline print-lisp.c]: ISO C forbids zero-size array ('auth_data').
tcpdump is build on many compilers != gcc

**Comment by fxlb** [comment]
More comments to come

**Comment by fxlb** [comment]
I have just updated `configure` to get more warnings on gcc-ism or non ISO 90 code, as we want to avoid them.
Please, rebase you branch on current upstream master and  `./configure; touch print-lisp.c; make -s`.
More comments to come.

**Comment by ritsz** [comment]
Rebased my commits, fixed the issues highlighted by you and the -Wpedantic issues with the new configure file. 
I'd like that "maps to" to be there. Highlights that the fact that the "<EID> maps to <LOC(s)>". Removed it for now in the latest commit.
Waiting for comments.

**Comment by fxlb** [comment]
Some comments and questions:
- "maps to" do not bring useful information. By RFC 6830, an EID maps to one or more RLOCs. A reader of the decoding is supposed to know that.
- Because details are often important, we really expect to have all fields/flags printed, possibly on multiple lines with indent.
- Some flags imply optional fields to be printed. I.e. in lisp_eid_register.pcap, Wireshark print xTR-ID and Site-ID.
- Which device built packets of the example pcap files?
- lisp_eid_notify.pcap: Wireshark print "Data" (24 bytes) on packet 2, do you know why?
- lisp_ipv6.pcap: Wireshark display "Malformed Packet" (pkt 1), do you know why?

**Comment by fxlb** [comment]
I suggest to put the "Map-Register message format" and the "Map-Notify message format" (from RFC) as comments to help in the understanding the code.

**Comment by ritsz** [comment]
I collected the pcaps on a Cisco CSR router. Copied the pcaps on a local machine and did a packet replay using scapy to test with tcpdump. While doing packet replay, I seem to have screwed up the I bit (xTR-ID present), which resulted in both the issues you saw in the pcaps. Issue would have been caught had I added the support for decoding all the flags. Adding them now, will do a fresh commit with the changed pcaps, printing flag, printing XTR-ID and Site-ID if I bit set.

**Comment by fxlb** [comment]
In such development, we have to think at the error cases, for example: the flag I or R is present, but not the corresponding field. Indeed, erroneous or malicious packets can appear.

**Comment by guyharris** [comment]
> In such development, we have to think at the error cases. For example: the flag I or R is present, but not the corresponding field. Indeed, erroneous or malicious packets can appear.

I.e., assume your dissection code will have thrown arbitrary random sequences of octets thrown at it; at some point, somebody probably *will* do so (or throw *targeted* random sequences of octets at it with, for example, [American Fuzzy Lop](http://lcamtuf.coredump.cx/afl/).

So don't discard the *un*-changed pcaps; make sure the printer handles them without, for example, overrunning the packet data, and keep both the unchanged ("broken") and changed pcaps as test cases.

**Comment by ritsz** [comment]
@guyharris @fxlb Added new test cases with a couple of malformed packets for EID Notify. Also printing new xTR-ID and SITE-ID if present and added packet structure as comments.

**Comment by fxlb** [comment]
Thanks for the updates.
I need a little time for a new review.

**Comment by fxlb** [comment]
- Currently, if the packet type is different from 'LISP_MAP_REGISTER' and 'LISP_MAP_NOTIFY', nothing is printed. It must print `LISP, <type_name>`.
- truncated versus malformed: Currently, if we replay the sample pcap files and capture with f.e. `tcpdump -s60 -vi lo`, all packets are printed ' [|MALFORMED LISP]'. By convention (see man page) `[|LISP]` is reserved for truncated packets (case we have a partial packet in the buffer). For undoubtedly malformed  packets, we could print `(malformed)`.
- For more legibility, many '3' and '4' (type) must be replaced by 'LISP_MAP_REGISTER' and 'LISP_MAP_NOTIFY'
- The field 'auth_data' is not printed
- can be record_count == 0 ? if no and it is : malformed.
- can be locator_count == 0 ? if no and it is : malformed.
-  simplify:
```
             if (loc_count > 1)
                        ND_PRINT((ndo, " %u Locators,", loc_count));
```
The test is useless, just use : ND_PRINT((ndo, " %u locator(s),", loc_count));
 (`s/%u Locators/%u locator(s)/`).
- lisp_eid_notify.pcap, packet 4: print 'Data' (24 bytes), What are these 'data'?
- The 'Register' in RFC6830 has ony 'P' and 'M' bits. Could you document the RFCs for 'S', 'I' and 'R' bits?
- The 'Notify' in RFC6830 has no header bits. Could you document the RFCs for 'I' and 'R' bits?
- Using the return value of `lisp_hdr_flag()` as xtr_present is unclear and using the value -1 to trigger return is tricky and does not seem future proof when other types will be decoded.
- For printing the flags, I should have been more specific: The usage is to print only the flags that are 'set'.
For this purpose, there is the bittok2str() function.
example:
If all P,S,I,R,M bits are valid for a Map-Register, we could use:
```
static const struct tok map_register_hdr_flag [] = {
       { 0x08000000, "P-Proxy-Map-Reply"  },
       { 0x04000000, "S-LISP-SEC-Capable" },
       { 0x02000000, "I-xTR-ID-Present"   },
       { 0x01000000, "R-Build-For-RTR"    },
       { 0x00000100, "M-Want-Map-Notify"  },
       { 0, NULL }
};

[...]

               ND_PRINT((ndo, "flags [%s]", bittok2str(map_register_hdr_flag,
                         "none", EXTRACT_32BITS(lisp_hdr))));
```
and get rid of some #define now useless.

**Comment by fxlb** [comment]
Same for other flags.

**Comment by fxlb** [comment]
Avoid blank lines in output.

**Comment by ritsz** [comment]
1. Prints the LISP type.
2. Malformed packets are printed as "(malformed-packet)".
3. Done.
4. Printing Auth data as hex data.
5. Record count cannot be zero, results in a malformed packet
6. Locator count can be zero, if the MS/MR has no record for the EID.
7. Done
8. This was a negetive test case. If the xTR_ID_Present bit is not set and yet the packet as additional data after all records are over, I am printing them as hex data.
9./10. Added a few comments to explain the use of these flags
11. Done
12. Done

**Comment by fxlb** [comment]
Thanks for the update. I need a little time for a new review.

**Comment by fxlb** [comment]
Thanks, the merge is close.
For a better history, I suggest you to  fold all your commits into one.
A way to do this: the `git rebase -i ...` command (man git-rebase ; INTERACTIVE MODE ; fixup command)
Furthermore, it is maybe simpler to close this pull request and open a new one.

**Comment by fxlb** [comment]
Previous comment updated.

**Comment by ritsz** [comment]
Closing this PR and opening a new one with rebased and squashed commits.

**Pull request closed without merging**